File: frontend/src/api/qemu.ts
/**
 * QEMU 虚拟机 API 端点
 * 提供 QEMU 虚拟机的完整生命周期管理功能
 */
import { get, post, put, del } from './request'
import type {
  QEMUVM,
  QEMUConfig,
  QEMUCreateParams,
  QEMUCloneParams,
  QEMUMigrateParams,
  QEMUSnapshot,
  CreateSnapshotParams,
  PendingConfig,
  RRDDataPoint,
  QueryOptions,
} from './types'

/**
 * 获取指定节点的 QEMU 虚拟机列表
 * @param node 节点名称
 * @param options 查询选项（包含 abort signal）
 */
export async function fetchQEMUList(
  node: string,
  options?: QueryOptions,
): Promise<QEMUVM[]> {
  return get<QEMUVM[]>(`/pve/nodes/${node}/qemu`, undefined, options)
}

/**
 * 获取 QEMU 虚拟机配置
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function getQEMUConfig(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<QEMUConfig> {
  return get<QEMUConfig>(`/pve/nodes/${node}/qemu/${vmid}/config`, undefined, options)
}

/**
 * 创建 QEMU 虚拟机
 * @param node 节点名称
 * @param params 创建参数
 * @param options 查询选项
 */
export async function createQEMU(
  node: string,
  params: QEMUCreateParams,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu`, params as Record<string, unknown>, options)
}

/**
 * 设置 QEMU 虚拟机配置
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param config 配置参数（键值对形式）
 * @param options 查询选项
 */
export async function setQEMUConfig(
  node: string,
  vmid: number,
  config: Record<string, unknown>,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/config`, config, options)
}

/**
 * 启动 QEMU 虚拟机
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function startQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/start`, undefined, options)
}

/**
 * 停止 QEMU 虚拟机（强制断电）
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function stopQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/stop`, undefined, options)
}

/**
 * 重启 QEMU 虚拟机
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function rebootQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/reboot`, undefined, options)
}

/**
 * 关闭 QEMU 虚拟机（通过 ACPI 优雅关机）
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function shutdownQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/shutdown`, undefined, options)
}

/**
 * 挂起 QEMU 虚拟机
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function suspendQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/suspend`, undefined, options)
}

/**
 * 恢复 QEMU 虚拟机（从挂起状态）
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function resumeQEMU(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(`/pve/nodes/${node}/qemu/${vmid}/status/resume`, undefined, options)
}

/**
 * 获取 QEMU 虚拟机快照列表
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function listSnapshots(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<QEMUSnapshot[]> {
  return get<QEMUSnapshot[]>(`/pve/nodes/${node}/qemu/${vmid}/snapshot`, undefined, options)
}

/**
 * 创建 QEMU 虚拟机快照
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param params 快照参数
 * @param options 查询选项
 */
export async function createSnapshot(
  node: string,
  vmid: number,
  params: CreateSnapshotParams,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(
    `/pve/nodes/${node}/qemu/${vmid}/snapshot`,
    params as Record<string, unknown>,
    options,
  )
}

/**
 * 删除 QEMU 虚拟机快照
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param snapname 快照名称
 * @param options 查询选项
 */
export async function deleteSnapshot(
  node: string,
  vmid: number,
  snapname: string,
  options?: QueryOptions,
): Promise<string> {
  return del<string>(`/pve/nodes/${node}/qemu/${vmid}/snapshot/${snapname}`, options)
}

/**
 * 回滚到指定快照
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param snapname 快照名称
 * @param options 查询选项
 */
export async function rollbackSnapshot(
  node: string,
  vmid: number,
  snapname: string,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(
    `/pve/nodes/${node}/qemu/${vmid}/snapshot/${snapname}/rollback`,
    undefined,
    options,
  )
}

/**
 * 克隆 QEMU 虚拟机
 * @param node 节点名称
 * @param vmid 源虚拟机 ID
 * @param params 克隆参数
 * @param options 查询选项
 */
export async function cloneQEMU(
  node: string,
  vmid: number,
  params: QEMUCloneParams,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(
    `/pve/nodes/${node}/qemu/${vmid}/clone`,
    params as Record<string, unknown>,
    options,
  )
}

/**
 * 迁移 QEMU 虚拟机到另一个节点
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param params 迁移参数
 * @param options 查询选项
 */
export async function migrateQEMU(
  node: string,
  vmid: number,
  params: QEMUMigrateParams,
  options?: QueryOptions,
): Promise<string> {
  return post<string>(
    `/pve/nodes/${node}/qemu/${vmid}/migrate`,
    params as Record<string, unknown>,
    options,
  )
}

/**
 * 获取 QEMU 虚拟机 RRD 监控数据
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param timeframe 时间范围 (hour|day|week|month|year)
 * @param dataset 数据集 (cpu|memory|network|disk|system)
 * @param options 查询选项
 */
export async function getQEMURRD(
  node: string,
  vmid: number,
  timeframe: string,
  dataset: string,
  options?: QueryOptions,
): Promise<RRDDataPoint[]> {
  return get<RRDDataPoint[]>(
    `/pve/nodes/${node}/qemu/${vmid}/rrd`,
    { timeframe, datasource: dataset },
    options,
  )
}

/**
 * 获取 QEMU 虚拟机待处理配置
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param options 查询选项
 */
export async function getQEMUPending(
  node: string,
  vmid: number,
  options?: QueryOptions,
): Promise<PendingConfig[]> {
  return get<PendingConfig[]>(
    `/pve/nodes/${node}/qemu/${vmid}/pending`,
    undefined,
    options,
  )
}

/**
 * 扩容 QEMU 虚拟机磁盘
 * @param node 节点名称
 * @param vmid 虚拟机 ID
 * @param disk 磁盘标识（如 scsi0、virtio0）
 * @param size 扩容大小（格式如 +10G）
 * @param options 查询选项
 */
export async function resizeQEMUDisk(
  node: string,
  vmid: number,
  disk: string,
  size: string,
  options?: QueryOptions,
): Promise<string> {
  return put<string>(
    `/pve/nodes/${node}/qemu/${vmid}/resize`,
    { disk, size } as Record<string, unknown>,
    options,
  )
}

File: backend/internal/pve/qemu.go
package pve

import (
	"context"
	"fmt"
	"net/url"
)

// ListQEMU 获取指定节点的 QEMU 虚拟机列表
// node: 节点名称
// 返回该节点上所有虚拟机的信息
func (c *Client) ListQEMU(ctx context.Context, node string) ([]QEMUVM, error) {
	var vms []QEMUVM
	path := fmt.Sprintf("nodes/%s/qemu", node)
	if err := c.Get(ctx, path, &vms); err != nil {
		return nil, fmt.Errorf("获取 QEMU 列表失败: %w", err)
	}
	return vms, nil
}

// GetQEMUConfig 获取指定虚拟机的完整配置
// node: 节点名称, vmid: 虚拟机 ID
// 返回虚拟机的详细配置参数
func (c *Client) GetQEMUConfig(ctx context.Context, node string, vmid int) (map[string]interface{}, error) {
	var config map[string]interface{}
	path := fmt.Sprintf("nodes/%s/qemu/%d/config", node, vmid)
	if err := c.Get(ctx, path, &config); err != nil {
		return nil, fmt.Errorf("获取 QEMU 配置失败: %w", err)
	}
	return config, nil
}

// CreateQEMU 创建新的 QEMU 虚拟机
// node: 节点名称, params: 创建参数
// 返回异步任务 ID (UPID)
func (c *Client) CreateQEMU(ctx context.Context, node string, params *QEMUCreateParams) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu", node)
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("创建 QEMU 失败: %w", err)
	}
	return upid, nil
}

// StartQEMU 启动虚拟机
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) StartQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/start", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("启动 QEMU 失败: %w", err)
	}
	return upid, nil
}

// StopQEMU 停止虚拟机（强制关机）
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) StopQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/stop", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("停止 QEMU 失败: %w", err)
	}
	return upid, nil
}

// ShutdownQEMU 关闭虚拟机（优雅关机，需要 QEMU guest agent）
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) ShutdownQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/shutdown", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("关闭 QEMU 失败: %w", err)
	}
	return upid, nil
}

// RebootQEMU 重启虚拟机
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) RebootQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/reboot", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("重启 QEMU 失败: %w", err)
	}
	return upid, nil
}

// SuspendQEMU 挂起虚拟机（保存状态到磁盘）
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) SuspendQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/suspend", node, vmid)
	params := map[string]interface{}{"todisk": 1}
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("挂起 QEMU 失败: %w", err)
	}
	return upid, nil
}

// ResumeQEMU 恢复虚拟机（从挂起状态恢复）
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) ResumeQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/resume", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("恢复 QEMU 失败: %w", err)
	}
	return upid, nil
}

// ResetQEMU 重置虚拟机（相当于断电重启）
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) ResetQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/reset", node, vmid)
	if err := c.Post(ctx, path, nil, &upid); err != nil {
		return "", fmt.Errorf("重置 QEMU 失败: %w", err)
	}
	return upid, nil
}

// SetQEMUConfig 更新虚拟机配置
// node: 节点名称, vmid: 虚拟机 ID, config: 配置参数映射
// 返回异步任务 ID (UPID)
func (c *Client) SetQEMUConfig(ctx context.Context, node string, vmid int, config QEMUConfigParams) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/config", node, vmid)
	if err := c.Put(ctx, path, config, &upid); err != nil {
		return "", fmt.Errorf("更新 QEMU 配置失败: %w", err)
	}
	return upid, nil
}

// ListQEMUSnapshots 获取虚拟机快照列表
// node: 节点名称, vmid: 虚拟机 ID
// 返回快照信息列表
func (c *Client) ListQEMUSnapshots(ctx context.Context, node string, vmid int) ([]Snapshot, error) {
	var snapshots []Snapshot
	path := fmt.Sprintf("nodes/%s/qemu/%d/snapshot", node, vmid)
	if err := c.Get(ctx, path, &snapshots); err != nil {
		return nil, fmt.Errorf("获取 QEMU 快照列表失败: %w", err)
	}
	return snapshots, nil
}

// CreateQEMUSnapshot 创建虚拟机快照
// node: 节点名称, vmid: 虚拟机 ID, name: 快照名称, description: 描述
// 返回异步任务 ID (UPID)
func (c *Client) CreateQEMUSnapshot(ctx context.Context, node string, vmid int, name, description string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/snapshot", node, vmid)
	params := url.Values{}
	params.Set("snapname", name)
	if description != "" {
		params.Set("description", description)
	}
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("创建 QEMU 快照失败: %w", err)
	}
	return upid, nil
}

// DeleteQEMUSnapshot 删除虚拟机快照
// node: 节点名称, vmid: 虚拟机 ID, snapname: 快照名称
// 返回异步任务 ID (UPID)
func (c *Client) DeleteQEMUSnapshot(ctx context.Context, node string, vmid int, snapname string) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/snapshot/%s", node, vmid, snapname)
	if err := c.Delete(ctx, path, &upid); err != nil {
		return "", fmt.Errorf("删除 QEMU 快照失败: %w", err)
	}
	return upid, nil
}

// CloneQEMU 克隆虚拟机
// node: 节点名称, vmid: 虚拟机 ID, params: 克隆参数
// 返回异步任务 ID (UPID)
func (c *Client) CloneQEMU(ctx context.Context, node string, vmid int, params *QEMUCloneParams) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/clone", node, vmid)
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("克隆 QEMU 失败: %w", err)
	}
	return upid, nil
}

// MigrateQEMU 迁移虚拟机到其他节点
// node: 节点名称, vmid: 虚拟机 ID, target: 目标节点名称
// 返回异步任务 ID (UPID)
func (c *Client) MigrateQEMU(ctx context.Context, node string, vmid int, params *QEMUMigrateParams) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d/migrate", node, vmid)
	if err := c.Post(ctx, path, params, &upid); err != nil {
		return "", fmt.Errorf("迁移 QEMU 失败: %w", err)
	}
	return upid, nil
}

// GetQEMURRD 获取虚拟机 RRD 性能数据
// node: 节点名称, vmid: 虚拟机 ID, timeframe: 时间范围 (hour, day, week, month, year)
// dataset: 数据集 (cpu, memory, network, disk)
func (c *Client) GetQEMURRD(ctx context.Context, node string, vmid int, timeframe, dataset string) ([]RRDPoint, error) {
	var data []RRDPoint
	path := fmt.Sprintf("nodes/%s/qemu/%d/rrd", node, vmid)
	params := url.Values{}
	params.Set("timeframe", timeframe)
	if dataset != "" {
		params.Set("ds", dataset)
	}
	if err := c.GetWithParams(ctx, path, params, &data); err != nil {
		return nil, fmt.Errorf("获取 QEMU RRD 数据失败: %w", err)
	}
	return data, nil
}

// GetQEMUPending 获取虚拟机待处理配置
// node: 节点名称, vmid: 虚拟机 ID
// 返回等待重启后生效的配置变更
func (c *Client) GetQEMUPending(ctx context.Context, node string, vmid int) ([]PendingConfig, error) {
	var pending []PendingConfig
	path := fmt.Sprintf("nodes/%s/qemu/%d/pending", node, vmid)
	if err := c.Get(ctx, path, &pending); err != nil {
		return nil, fmt.Errorf("获取 QEMU 待处理配置失败: %w", err)
	}
	return pending, nil
}

// DeleteQEMU 删除虚拟机
// node: 节点名称, vmid: 虚拟机 ID
// 返回异步任务 ID (UPID)
func (c *Client) DeleteQEMU(ctx context.Context, node string, vmid int) (string, error) {
	var upid string
	path := fmt.Sprintf("nodes/%s/qemu/%d", node, vmid)
	if err := c.Delete(ctx, path, &upid); err != nil {
		return "", fmt.Errorf("删除 QEMU 失败: %w", err)
	}
	return upid, nil
}

// GetQEMUCurrent 获取虚拟机当前状态
// node: 节点名称, vmid: 虚拟机 ID
// 返回虚拟机当前运行状态
func (c *Client) GetQEMUCurrent(ctx context.Context, node string, vmid int) (*QEMUVM, error) {
	var vm QEMUVM
	path := fmt.Sprintf("nodes/%s/qemu/%d/status/current", node, vmid)
	if err := c.Get(ctx, path, &vm); err != nil {
		return nil, fmt.Errorf("获取 QEMU 当前状态失败: %w", err)
	}
	return &vm, nil
}
